scipy/optimize/_linesearch.py

```diff
@@ -725,7 +725,10 @@ def scalar_search_armijo(phi, phi0, derphi0, c1=1e-4, alpha0=1, amin=0):
         phi_a2 = phi(alpha2)
 
         if (phi_a2 <= phi0 + c1*alpha2*derphi0):
-            return alpha2, phi_a2
+            if (alpha2 > amin):
+                return alpha2, phi_a2
+            else:
+                return alpha1, phi_a1
```

Comment:
Sorry this has taken so long, I wanted to confirm why it's ok to return `alpha1` when it doesn't satisfy armijio?

Reply:
As far as I'm aware, there is no guarantee that there exists an `alpha` satisfying armijo. The proposed change simply makes sure that the returned value satisfies `alpha >= amin` (a requirement of the function that was violated). The PR proposes a **minimal change** to ensure `alpha >= amin` in this case, but does not pretend to be exhaustive. The test I provided highlights this fact (in both cases, there can be cases where  `alpha < amin`).

TBH, I think functional specs need to be decided upon concerning `scalar_search_armijo` because it is not clear to me what condition should take precedence in case of clash.

Reply:
Thanks for explaining. One thing that makes me a bit uneasy is that the while loop has the same condition after which 
```
# Failed to find a suitable step length
 return None, phi_a1
```
So I was wondering if this should also fail by returning `None` too?

Reply:
> So I was wondering if this should also fail by returning None too?

@gabrielfougeron would you be able to respond to this review comment?

Reply:
Hey, sorry for my delayed response.

@j-bowhay : Long story short, I don't know.

IMO, what this whole section of the code really needs is a set of motivated functional design specifications. Why am I saying that ? On the one hand, YES, this kind of function should do what they advertise. On the other hand, they are not part of the public API, and "what they advertise" might not be the best strategy in the context of optimization / root search.

I think your comment hits the nail right on the head. It is a widespread pattern on the lower levels of scipy.optimize to have (often quite good) reference implementations of classical algorithms, that kind of misses the bigger picture because  it is **extremely hard** to anticipate how the different procedures will behave together (example: Newton + inexact krylov solver (e.g. LGMRES) + linesearch).

The reason why I created the pull request was to stop the behavior "returned alpha is always very small", which makes the whole Newton search stumble.